federation-js/src/composition/composeAndValidate.ts

```diff
@@ -7,12 +7,38 @@ import {
 import { ServiceDefinition } from './types';
 import { normalizeTypeDefs } from './normalize';
 import { printComposedSdl } from '../service/printComposedSdl';
+import { GraphQLError, GraphQLSchema } from 'graphql';
 
-export function composeAndValidate(serviceList: ServiceDefinition[]) {
+export type CompositionResult = CompositionFailure | CompositionSuccess;
+
+// Yes, it's a bit awkward that we still return a schema when errors occur.
+// This is old behavior that I'm choosing not to modify for now.
+export interface CompositionFailure {
+  /** @deprecated Use composedSdl instead */
+  schema: GraphQLSchema;
+  errors: GraphQLError[];
+}
+
+export interface CompositionSuccess {
+  /** @deprecated Use composedSdl instead */
+  schema: GraphQLSchema;
+  composedSdl: string;
+}
```

Comment:
Included some type refinement work in this PR. Handling of composition results is a bit more strictly typed now. You either have `errors` _or_ `composedSdl`, which is now enforced by the type system itself.